File: python/socketd/socketd/transport/client/ClientBase.py
from abc import ABC, abstractmethod
from typing import Optional

from socketd.exception.SocketDExecption import SocketDException
from socketd.transport.client.Client import ClientInternal, Client
from socketd.transport.client.ClientChannel import ClientChannel
from socketd.transport.client.ClientConfigHandler import ClientConfigHandler
from socketd.transport.client.ClientConnectHandler import ClientConnectHandler, ClientConnectHandlerDefault
from socketd.transport.client.ClientConnector import ClientConnector
from socketd.transport.core.ChannelAssistant import ChannelAssistant
from socketd.transport.core.Costants import Constants
from socketd.transport.core.Session import Session
from socketd.transport.client.ClientHeartbeatHandler import ClientHeartbeatHandler
from socketd.utils.LogConfig import log
from socketd.transport.core.impl.ProcessorDefault import ProcessorDefault
from socketd.transport.client.ClientConfig import ClientConfig



class ClientBase(ClientInternal, ABC):

    def __init__(self, client_config: ClientConfig, assistant: ChannelAssistant):
        self.__config: ClientConfig = client_config
        self.__assistant: ChannelAssistant = assistant

        self.__processor = ProcessorDefault()
        self.__heartbeat_handler: Optional[ClientHeartbeatHandler] = None
        self.__connect_handler: type[ClientConnectHandler] = ClientConnectHandlerDefault

    def get_assistant(self):
        return self.__assistant

    def get_config(self) -> ClientConfig:
        return self.__config

    def get_processor(self):
        return self.__processor

    def get_connect_handler(self) -> ClientConnectHandler:
        return self.__connect_handler

    def get_heartbeat_interval(self) -> int:
        return self.__config.get_heartbeat_interval()

    def get_heartbeat_handler(self) -> ClientHeartbeatHandler:
        return self.__heartbeat_handler

    def connect_handler(self, connectHandler: ClientConnectHandler) -> Client:
        if connectHandler:
            self.__connect_handler = connectHandler
        return self

    def heartbeat_handler(self, heartbeatHandler: ClientHeartbeatHandler):
        if heartbeatHandler:
            self.__heartbeat_handler = heartbeatHandler
        return self

    def config(self, configHandler: ClientConfigHandler):
        if configHandler:
            configHandler(self.__config)
        return self

    def process(self, processor):
        if processor:
            self.__processor = processor
        return self

    def listen(self, listener):
        if listener:
            self.__processor.set_listener(listener)
        return self

    async def _open_do(self, isThrow):
        connector: ClientConnector = self.create_connector()
        clientChannel: ClientChannel = ClientChannel(self, connector)

        try:
            await clientChannel.connect()
            log.info(f"Socket.D client successfully connected: link={self.get_config().get_link_url()}")
        except Exception as e:
            if isThrow:
                await clientChannel.close(code=Constants.CLOSE2008_OPEN_FAIL)
                raise SocketDException(f"Socket.D client Connection failed {e}")
            else:
                log.info(f"Socket.D client Connection failed: link={self.get_config().get_link_url()}")

        return clientChannel.get_session()

    async def open(self) -> Session:
        return await self._open_do(False)

    async def open_or_throw(self) -> Session:
        return await self._open_do(True)

    @abstractmethod
    def create_connector(self):
        ...

File: javascript/socketd/src/socketd/transport/client/ClientBase.ts
import type {ChannelAssistant} from "../core/ChannelAssistant";
import type {ClientConfig} from "./ClientConfig";
import type {IoConsumer, IoFunction} from "../core/Typealias";
import type {ClientConnector} from "./ClientConnector";
import {ChannelInternal} from "../core/Channel";
import type {Session} from "../core/Session";
import {Processor} from "../core/Processor";
import {ProcessorDefault} from "../core/impl/ProcessorDefault";
import type {Listener} from "../core/Listener";
import type {ClientSession} from "./ClientSession";
import {ClientChannel} from "./ClientChannel";
import {Constants} from "../core/Constants";
import {Client, ClientInternal} from "./Client";

/**
 * 客户端基类
 *
 * @author noear
 * @since 2.0
 */
export abstract class ClientBase<T extends ChannelAssistant<Object>> implements ClientInternal {
    private _config: ClientConfig;
    private _connectHandler : IoFunction<ClientConnector, Promise<ChannelInternal>>;
    private _heartbeatHandler: IoConsumer<Session>;
    private _processor: Processor;
    private _assistant: T;

    constructor(clientConfig: ClientConfig, assistant: T) {
        this._config = clientConfig;
        this._assistant = assistant;
        this._processor = new ProcessorDefault();
    }

    /**
     * 获取通道助理
     */
    getAssistant(): T {
        return this._assistant;
    }

    getConnectHandler(): IoFunction<ClientConnector, Promise<ChannelInternal>> {
        return this._connectHandler;
    }

    /**
     * 获取心跳处理
     */
    getHeartbeatHandler(): IoConsumer<Session> {
        return this._heartbeatHandler;
    }

    /**
     * 获取心跳间隔（毫秒）
     */
    getHeartbeatInterval(): number {
        return this.getConfig().getHeartbeatInterval();
    }

    /**
     * 获取配置
     */
    getConfig(): ClientConfig {
        return this._config;
    }

    /**
     * 获取处理器
     */
    getProcessor(): Processor {
        return this._processor;
    }

    /**
     * 设置连接处理器
     */
    connectHandler(handler: IoFunction<ClientConnector, Promise<ChannelInternal>>) {
        if (handler != null) {
            this._connectHandler = handler;
        }

        return this;
    }

    /**
     * 设置心跳处理器
     */
    heartbeatHandler(handler: IoConsumer<Session>) {
        if (handler != null) {
            this._heartbeatHandler = handler;
        }

        return this;
    }

    /**
     * 配置
     */
    config(configHandler: IoConsumer<ClientConfig>) {
        if (configHandler != null) {
            configHandler(this._config);
        }

        return this;
    }

    /**
     * 设置监听器
     */
    listen(listener: Listener): Client {
        if (listener != null) {
            this._processor.setListener(listener);
        }

        return this;
    }


    open(): Promise<ClientSession> {
        return this.openDo(false);
    }

    /**
     * 打开会话
     */
    openOrThow(): Promise<ClientSession> {
        return this.openDo(true);
    }

    private openDo(isThow: boolean): Promise<ClientSession> {
        const connector = this.createConnector();
        const clientChannel = new ClientChannel(this, connector);

        return new Promise<ClientSession>((resolve, reject) => {
            // @ts-ignore
            clientChannel.connect().then(res => {
                console.info("Socket.D client successfully connected!");
                resolve(clientChannel.getSession());
            }, err => {
                if (isThow) {
                    clientChannel.close(Constants.CLOSE2008_OPEN_FAIL);
                    reject(err);
                } else {
                    console.warn("Socket.D client Connection failed!");
                    resolve(clientChannel.getSession());
                }
            })
        });
    }

    /**
     * 创建连接器
     */
    protected abstract createConnector(): ClientConnector;
}
